### ISSUE TITLE
doesnt work in SDwebui

### ISSUE THREAD

**Issue by JiliHili** [THUDM/ImageReward#15]
Error completing request
Arguments: (0, '{nynaneki _ 1.5},The Petite gril with Ocean Camo  Coat on the street,masterpiece_1.1, coloerful cloud,bloom,fantasy,sunlight,sun', 'sea', 'None', 'None', <PIL.Image.Image image mode=RGB size=1920x1440 at 0x1F3F4613100>, None, None, None, 0, 20, 0, 4, 1, False, False, 1, 1, 7, 0.75, -1.0, -1.0, 0, 0, 0, False, 512, 512, 0, False, 32, 0, '', '', 1, '', '', 1, 50, 0, False, 4, 1, '<p style="margin-bottom:0.75em">Recommended settings: Sampling Steps: 80-100, Sampler: Euler a, Denoising strength: 0.8</p>', 128, 8, ['left', 'right', 'up', 'down'], 1, 0.05, 128, 4, 0, ['left', 'right', 'up', 'down'], False, False, None, '', '<p style="margin-bottom:0.75em">Will upscale the image to twice the dimensions; use width and height sliders to set tile size</p>', 64, 0, 1, '', 0, '', True, False) {}
Traceback (most recent call last):
  File "D:\Ts\modules\ui.py", line 184, in f
    res = list(func(*args, **kwargs))
  File "D:\Ts\webui.py", line 64, in f
    res = func(*args, **kwargs)
  File "D:\Ts\modules\img2img.py", line 124, in img2img
    processed = modules.scripts.scripts_img2img.run(p, *args)
  File "D:\Ts\modules\scripts.py", line 162, in run
    processed = script.run(p, *script_args)
TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'

**Comment by tongyx361** [comment]
Could you please provide us with some information to reproduce this error?

**Comment by JiliHili** [comment]
> Thank you so much for the issue! Let's resolve it together 😀
> 
> Sorry that I failed to reproduce the error. Could you please provide us with more information such as generation configs, SD Web UI settings, command line args?

I install it with pip command
and i download script from github
![QQ截图20230511071614](https://github.com/THUDM/ImageReward/assets/102945328/468808f0-5588-49a8-9416-a7e3aa216fac)
![QQ截图20230511072545](https://github.com/THUDM/ImageReward/assets/102945328/3d65a751-9902-462a-b960-ed3cd46f773f)
this is my terminal print

```
venv "D:\Ts\venv\Scripts\Python.exe"
Python 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
Commit hash: 04c0e643f2eec68d93a76db171b4d70595808702
Launching Web UI with arguments: --lowvram --disable-safe-unpickle --xformers --deepdanbooru --no-half-vae
Cannot import xformers
Traceback (most recent call last):
  File "D:\Ts\modules\sd_hijack_optimizations.py", line 18, in <module>
    import xformers.ops
ModuleNotFoundError: No module named 'xformers'

LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
making attention of type 'vanilla' with 512 in_channels
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla' with 512 in_channels
Loading weights [c3841625] from D:\Ts\models\Stable-diffusion\model.ckpt
Applying cross attention optimization (Doggettx).
Model loaded.
**Error calling: D:\Ts\scripts\image-reward.py/ui
Traceback (most recent call last):
  File "D:\Ts\modules\scripts.py", line 87, in wrap_call
    res = func(*args, **kwargs)
  File "D:\Ts\scripts\image-reward.py", line 50, in ui
    unload_button.click(unload_image_reward_model)
TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'

Error calling: D:\Ts\scripts\image-reward.py/ui
Traceback (most recent call last):
  File "D:\Ts\modules\scripts.py", line 87, in wrap_call
    res = func(*args, **kwargs)
  File "D:\Ts\scripts\image-reward.py", line 50, in ui
    unload_button.click(unload_image_reward_model)
TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'**

Loaded a total of 1 textual inversion embeddings.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
```

**Comment by tongyx361** [comment]
> > Thank you so much for the issue! Let's resolve it together 😀
> > Sorry that I failed to reproduce the error. Could you please provide us with more information such as generation configs, SD Web UI settings, command line args?
> 
> I install it with pip command and i download script(demo/sdwebui/image-reward.py) from github ![QQ截图20230511071614](https://user-images.githubusercontent.com/102945328/237534915-468808f0-5588-49a8-9416-a7e3aa216fac.png) ![QQ截图20230511072545](https://user-images.githubusercontent.com/102945328/237535236-3d65a751-9902-462a-b960-ed3cd46f773f.png) this is my terminal print
> 
> ```
> venv "D:\Ts\venv\Scripts\Python.exe"
> Python 3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]
> Commit hash: 04c0e643f2eec68d93a76db171b4d70595808702
> Launching Web UI with arguments: --lowvram --disable-safe-unpickle --xformers --deepdanbooru --no-half-vae
> Cannot import xformers
> Traceback (most recent call last):
>   File "D:\Ts\modules\sd_hijack_optimizations.py", line 18, in <module>
>     import xformers.ops
> ModuleNotFoundError: No module named 'xformers'
> 
> LatentDiffusion: Running in eps-prediction mode
> DiffusionWrapper has 859.52 M params.
> making attention of type 'vanilla' with 512 in_channels
> Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
> making attention of type 'vanilla' with 512 in_channels
> Loading weights [c3841625] from D:\Ts\models\Stable-diffusion\model.ckpt
> Applying cross attention optimization (Doggettx).
> Model loaded.
> **Error calling: D:\Ts\scripts\image-reward.py/ui
> Traceback (most recent call last):
>   File "D:\Ts\modules\scripts.py", line 87, in wrap_call
>     res = func(*args, **kwargs)
>   File "D:\Ts\scripts\image-reward.py", line 50, in ui
>     unload_button.click(unload_image_reward_model)
> TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'
> 
> Error calling: D:\Ts\scripts\image-reward.py/ui
> Traceback (most recent call last):
>   File "D:\Ts\modules\scripts.py", line 87, in wrap_call
>     res = func(*args, **kwargs)
>   File "D:\Ts\scripts\image-reward.py", line 50, in ui
>     unload_button.click(unload_image_reward_model)
> TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'**
> 
> Loaded a total of 1 textual inversion embeddings.
> Running on local URL:  http://127.0.0.1:7860
> 
> To create a public link, set `share=True` in `launch()`.
> ```

# 中文版

看起来你是一名中国同胞，所以在接下来的 thread 中，我会提供中文回复，同时考虑到 GitHub 是一个国际化的开源社区，我也会保留英文回复，以飨遇到类似问题但来自其他国家的用户。

请尽管根据你的喜好选择使用中文或英文来阅读或回复。

## 简单总结

1. 为了解决 `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'`，尝试尝试将你的 `image-reward.py` 中的
    ```python
    unload_button.click(unload_image_reward_model)
    ```
    **替换**为
    ```python
    unload_button.click(unload_image_reward_model, inputs=[], outputs=[])
    ```
2. 为了解决 `TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'`，可能需要你进一步提供**引发这一错误时的 SD Web UI 的整个界面截图**。

## 详细内容

### 关于 `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'` indicates that the `click()`

首先，你这条信息中提供的终端打印信息中报告的错误与该 issue 一开始报告的错误不同。
我们先来解决这条信息中的 `image-reward.py` 引发的错误。
`TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'` 表示 `click()` 函数缺失了两个参数 `inputs` 和 `outputs`。
实际上，没有这两个参数是我们的刻意设计，因为这一点击事件不需要任何输入或输出，而且在我们的 Python3.10.6 的测试环境中这也不会引发任何错误。

但无论如何，为了解决这一问题，你可以尝试将你的 `image-reward.py` 中的
```python
unload_button.click(unload_image_reward_model)
```
替换为
```python
unload_button.click(unload_image_reward_model, inputs=[], outputs=[])
```

### 关于 `TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'`

其次，我仍然无法根据当前已有的信息复现该 issue 一开始报告的错误，即 `TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'`。
我认为对于复现这一错误，比较重要的信息应该是你在生成图片时所使用的设置。
能否将你引发这一错误时的 SD Web UI 的整个界面截图并发送给我？

### 关于 `ModuleNotFoundError: No module named 'xformers'` 的附带提示

附带地，你的终端报告信息中 `ModuleNotFoundError: No module named 'xformers'` 表示你没有安装 `xformers`，但在启动 SD Web UI 时使用了 `--xformers` 命令行参数。如果你不打算安装并使用 `xformers`，可以考虑在以后启动时将 `--xformers` 命令行参数去掉。

# English ver.

It looks like you're also Chinese, so in the following thread, I'll be providing replies in Chinese.
However, since GitHub is an international open source community, I will also provide English responses in the good of users from other country with similar problems.

Please fell free to read or respond in English or Chinese, depending on your preference.

## TL;DR

1. To solve the error `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'`, try replacing
    ```python
    unload_button.click(unload_image_reward_model)
    ```
    with
    ```python
    unload_button.click(unload_image_reward_model, inputs=[], outputs=[])
    ```
2. To solve the error `TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'`, you may need to provide **a screenshot of the entire interface of SD Web UI when the error occurred**.

## Details

### About `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'` indicates that the `click()`

Firstly, the error reported in your terminal output is different from the original error reported in this issue. Let's first solve the error caused by `image-reward.py`.
The `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'` indicates that the `click()` function is missing two parameters, `inputs` and `outputs`, which, In fact, are intentionally not included, as this click event does not require any input or output, and in our testing environment with Python 3.10.6, this does not cause any errors.

But anyway, to fix this, you can try replacing
  ```python
  unload_button.click(unload_image_reward_model)
  ```
with
  ```python
  unload_button.click(unload_image_reward_model, inputs=[], outputs=[])
  ```

### About `TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'`

Secondly, I am still unable to reproduce the original error reported in this issue, namely `TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'`.
I believe that the important information for reproducing this error should be the settings you used when generating the image.
Could you please provide me with a screenshot of the entire interface of SD Web UI when the error occurred?

### Incidental Tip about `ModuleNotFoundError: No module named 'xformers'`

Incidentally, the `ModuleNotFoundError: No module named 'xformers'` in your terminal output indicates that you have not installed `xformers`, but you used the `--xformers` command line parameter when launching the SD Web UI. If you do not plan to install and use `xformers`, you can consider removing the `--xformers` command line parameter when launching SD Web UI in the future.

**Comment by JiliHili** [comment]
TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'这个是在ui里点击generate造成的
TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'是启动的报错
改完脚本以后这两个报错变成了另一个
**Traceback (most recent call last):
  File "D:\Ts\scripts\image-reward.py", line 56, in run
    shared.image_reward_model  # if loaded, do nothing
AttributeError: module 'modules.shared' has no attribute 'image_reward_model'**
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\Ts\modules\ui.py", line 184, in f
    res = list(func(*args, **kwargs))
  File "D:\Ts\webui.py", line 64, in f
    res = func(*args, **kwargs)
  File "D:\Ts\modules\txt2img.py", line 40, in txt2img
    processed = modules.scripts.scripts_txt2img.run(p, *args)
  File "D:\Ts\modules\scripts.py", line 162, in run
    processed = script.run(p, *script_args)
  File "D:\Ts\scripts\image-reward.py", line 63, in run
    shared.image_reward_model = reward.load(
  File "D:\Ts\venv\lib\site-packages\ImageReward\utils.py", line 65, in load
    model_path = ImageReward_download(_MODELS[name], download_root or os.path.expanduser("~/.cache/ImageReward"))
  File "D:\Ts\venv\lib\site-packages\ImageReward\utils.py", line 41, in ImageReward_download
    hf_hub_download(repo_id="THUDM/ImageReward", filename=filename, local_dir=root)
  File "D:\Ts\venv\lib\site-packages\huggingface_hub\utils\_validators.py", line 120, in _inner_fn
    return fn(*args, **kwargs)
  File "D:\Ts\venv\lib\site-packages\huggingface_hub\file_download.py", line 1286, in hf_hub_download
    return _to_local_dir(blob_path, local_dir, relative_filename, use_symlinks=local_dir_use_symlinks)
  File "D:\Ts\venv\lib\site-packages\huggingface_hub\file_download.py", line 1594, in _to_local_dir
    _create_symlink(real_blob_path, local_dir_filepath, new_blob=False)
  File "D:\Ts\venv\lib\site-packages\huggingface_hub\file_download.py", line 873, in _create_symlink
    _support_symlinks = are_symlinks_supported(os.path.dirname(os.path.commonpath([abs_src, abs_dst])))
  File "D:\PHON\Lib\ntpath.py", line 763, in commonpath
    raise ValueError("Paths don't have the same drive")
ValueError: Paths don't have the same drive
![QQ截图20230511145500](https://github.com/THUDM/ImageReward/assets/102945328/e92cd9b5-8d88-4fa8-9da8-0786117e11ec)

**Comment by tongyx361** [comment]
> TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'这个是在ui里点击generate造成的 TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'是启动的报错 改完脚本以后这两个报错变成了另一个 **Traceback (most recent call last): File "D:\Ts\scripts\image-reward.py", line 56, in run shared.image_reward_model # if loaded, do nothing AttributeError: module 'modules.shared' has no attribute 'image_reward_model'** During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last): File "D:\Ts\modules\ui.py", line 184, in f res = list(func(*args, **kwargs)) File "D:\Ts\webui.py", line 64, in f res = func(*args, **kwargs) File "D:\Ts\modules\txt2img.py", line 40, in txt2img processed = modules.scripts.scripts_txt2img.run(p, *args) File "D:\Ts\modules\scripts.py", line 162, in run processed = script.run(p, *script_args) File "D:\Ts\scripts\image-reward.py", line 63, in run shared.image_reward_model = reward.load( File "D:\Ts\venv\lib\site-packages\ImageReward\utils.py", line 65, in load model_path = ImageReward_download(_MODELS[name], download_root or os.path.expanduser("~/.cache/ImageReward")) File "D:\Ts\venv\lib\site-packages\ImageReward\utils.py", line 41, in ImageReward_download hf_hub_download(repo_id="THUDM/ImageReward", filename=filename, local_dir=root) File "D:\Ts\venv\lib\site-packages\huggingface_hub\utils_validators.py", line 120, in _inner_fn return fn(*args, **kwargs) File "D:\Ts\venv\lib\site-packages\huggingface_hub\file_download.py", line 1286, in hf_hub_download return _to_local_dir(blob_path, local_dir, relative_filename, use_symlinks=local_dir_use_symlinks) File "D:\Ts\venv\lib\site-packages\huggingface_hub\file_download.py", line 1594, in _to_local_dir _create_symlink(real_blob_path, local_dir_filepath, new_blob=False) File "D:\Ts\venv\lib\site-packages\huggingface_hub\file_download.py", line 873, in _create_symlink _support_symlinks = are_symlinks_supported(os.path.dirname(os.path.commonpath([abs_src, abs_dst]))) File "D:\PHON\Lib\ntpath.py", line 763, in commonpath raise ValueError("Paths don't have the same drive") ValueError: Paths don't have the same drive ![QQ截图20230511145500](https://user-images.githubusercontent.com/102945328/237598310-e92cd9b5-8d88-4fa8-9da8-0786117e11ec.png)

> # Translation to English for Reference (after Fixing an Error)
> `TypeError: run() missing 2 required positional arguments: 'filter_out_low_scores' and 'lower_score_limit'` caused by clicking "Generate" in the UI
> `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'` is the launching error.
> After changing the script, these two errors have become another one. 
> (Same below.)

# 中文版

你这条消息开头提到的两个错误是相同的，显然不符合事实。我猜，其中第二个错误本来应该是 `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'`？

## 简单总结

对于你（以及其他遇到类似问题的 Windows 系统用户），如果你仅希望在 SD Web UI 中使用 ImageReward，那么你可以将选择 ImageReward 安装到 SD Web UI 的 `models` 目录中。具体而言，你需要在你的 `image-reward.py` 中将
```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
)  # using shared to make the model object global among modules
```
替换为
```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
    download_root=Path.cwd().parent / "models" / "ImageReward",
)  # using shared to make the model object global among modules
```

如果你希望在 SD Web UI 外的其他地方使用 ImageReward，欢迎进一步咨询。

## 细节说明

你目前的问题是 `ValueError: Paths don't have the same drive` 导致了 ImageReward 安装失败。

`ValueError: Paths don't have the same drive` 表示 （Windows 系统中）至少有两个路径所在的硬盘分区不同（例如一个路径以 "C:" 开头，另一个路径以 "D:" 开头）。

此处引发错误的两个路径分别是 Hugging Face 的 `cache_dir` 与 `local_dir`。

ImageReward 默认从 Hugging Face Hub 下载模型，而 `cache_dir` 与 `local_dir` 正是 Hugging Face 下载模型涉及到的两个路径。
Hugging Face 下载模型的方式是先将所有文件下载到 `cache_dir` 内，再视文件大小在 `local_dir` 内创建到 `cache_dir` 内大文件的链接/直接将`cache_dir` 内的小文件复制到 `local_dir`，最终 Hugging Face 默认从 `local_dir` 中调用模型。

ImageReward 默认将 `local_dir` 设置在本地用户目录中的 `~/.cache/ImageReward`，以便用户在其他地方使用 ImageReward 时无需再次下载。
在 Windows 系统中，`~/.cache/ImageReward` 的 `~` 一般被转化为 C 盘中的用户目录；
此时，如果你的 `cache_dir` 被设置为其他盘（例如 D 盘）中的目录路径，
那么，`local_dir` 与 `cache_dir` 就位于不同盘中，进而引发上述错误。

而 `cache_dir` 所在的目录默认由环境变量 `HF_HOME` 设置。
你可以在 cmd 中输入下述命令，检查你的 `HF_HOME` 值是否确实不在 C 盘中。
```shell
echo %HF_HOME%
```

因此，对于你（以及其他遇到类似问题的 Windows 系统用户），一种可能的解决方法是：如果你仅希望在 SD Web UI 中使用 ImageReward，那么你可以将选择 ImageReward 安装到 SD Web UI 的 `models` 目录中。具体而言，你需要在你的 `image-reward.py` 中将

```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
)  # using shared to make the model object global among modules
```

替换为

```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
    download_root=Path.cwd().parent / "models" / "ImageReward",
)  # using shared to make the model object global among modules
```

如果你希望在 SD Web UI 外的其他地方使用 ImageReward，欢迎进一步咨询。

# English ver.

The two errors you mentioned at the beginning of this message are the same, which obviously makes no sense. I guess the second of these errors should have been `TypeError: click() missing 2 required positional arguments: 'inputs' and 'outputs'`?

## TL;DR

A possible workaround for you (and other Windows users experiencing similar problems) is that: if you only wish to use ImageReward in SD Web UI, then you can choose to install the ImageReward into the `models` directory of SD Web UI. Specifically, in your `image-reward.py`, you need to replace the

```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
) # using shared to make the model object global among modules
```

with

```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
    download_root=Path.cwd().parent / "models" / "ImageReward",
) # using shared to make the model object global among modules
```

If you wish to use ImageReward elsewhere than in the SD Web UI, feel free to inquire further.

## Details

Your current problem is that `ValueError: Paths don't have the same drive` caused the ImageReward installation to fail.

`ValueError: Paths don't have the same drive` means that at least two paths (on Windows systems) are on different hard drive partitions (e.g. one path starts with "C:" and the other path starts with "D:").

The two paths that cause the error here are `cache_dir` and `local_dir` in Hugging Face.

ImageReward downloads models from the Hugging Face Hub by default, and `cache_dir` and `local_dir` are the two paths involved in downloading models from the Hugging Face Hub.
Hugging Face downloads the model by first downloading all files to `cache_dir`, then creating links in `local_dir` to large files in `cache_dir` depending on the file size / copying small files in `cache_dir` directly to `local_dir`, and finally Hugging Face Call the model from `local_dir` by default.

ImageReward sets `local_dir` to `~/.cache/ImageReward` in the local user directory by default, so that users do not have to download it again when using ImageReward elsewhere.
On Windows systems, `~` in `~/.cache/ImageReward` is generally translated to the user directory on the C drive;
At this point, if your `cache_dir` is set to a directory path in another drive (e.g. D drive),
`local_dir` and `cache_dir` will be on different disks, which will cause the above error.

The directory where `cache_dir` is located is set by the environment variable `HF_HOME` by default.
You can check if your `HF_HOME` value is not in the C drive by typing the following command in cmd.
```shell
echo %HF_HOME%
```

So, a possible workaround for you (and other Windows users experiencing similar problems) is that: if you only wish to use ImageReward in SD Web UI, you can choose to install the ImageReward into the `models` directory of SD Web UI. Specifically, in your `image-reward.py`, you need to replace the

```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
) # using shared to make the model object global among modules
```

with

```Python
shared.image_reward_model = reward.load(
    "ImageReward-v1.0",
    download_root=Path.cwd().parent / "models" / "ImageReward",
) # using shared to make the model object global among modules
```

If you wish to use ImageReward in other places outside of the SD Web UI, feel free to inquire further.
